src/parser/symtab.ts

```diff
@@ -0,0 +1,1738 @@
+       This is independent from how the name is used. */
+    if (privateobj === null /* !PyUnicode_Check(privateobj) || */ || !ident.startsWith("__")) {
+        return ident;
+    }
+    /* Don't mangle __id__ or names with dots.
+          The only time a name with a dot can occur is when
+          we are compiling an import statement that has a
+          package name.
+          TODO(jhylton): Decide whether we want to support
+          mangling of the module name, e.g. __M.X.
+       */
+    if (ident.endsWith("__") || ident.indexOf(".") != -1) {
+        return ident; /* Don't mangle __whatever__ */
+    }
+
+    /* Strip leading underscores from class name */
+    const stripped = privateobj.replace(/^_+/, "");
+    if (stripped.length === 0) {
+        return ident; /* Don't mangle if class is just underscores */
+    }
+
+    /* ident = "_" + priv[ipriv:] + ident # i.e. 1+plen+nlen bytes */
+    return `_${stripped}${ident}`;
+}
+
+export class SymbolTable {
+    filename: string;
+    cur: SymbolTableScope | null = null;
+    top: SymbolTableScope | null = null;
+    stack: SymbolTableScope[] = [];
+    global: { [name: string]: number } = {};
+    curClass: string | null = null;
+    tmpname = 0;
+
+    stss: { [thing: number]: SymbolTableScope } = {};
+
+    constructor(filename: string, future: never) {
+        this.filename = filename;
+    }
+
+    get private() {
+        return this.curClass;
+    }
+
+    getStsForAst(ast: AST) {
+        var v;
+        assert(ast.scopeId !== undefined, "ast wasn't added to st?");
+        v = this.stss[ast.scopeId];
+        assert(v !== undefined, "unknown sym tab entry");
+        return v;
+    }
+
+    addDef(name: string, flag: SYMTAB_CONSTS, ste: SymbolTableScope | null = null) {
+        const cur_ste = ste || this.cur;
+        assert(cur_ste !== null, "need to know what we're operating on");
+
+        const mangled = _Py_Mangle(this.private, name);
+        let val = cur_ste.symbols[mangled];
+        if (val !== undefined) {
+            if (flag & SYMTAB_CONSTS.DEF_PARAM && val & SYMTAB_CONSTS.DEF_PARAM) {
+                throw new pySyntaxError("duplicate argument '" + name + "' in function definition", [
+                    this.filename,
+                    cur_ste.lineno,
+                    cur_ste.colOffset,
+                    "",
+                ]);
+            }
+            val |= flag;
+        } else {
+            val = flag;
+        }
+
+        if (cur_ste.comp_iter_target) {
+            /* This name is an iteration variable in a comprehension,
+             * so check for a binding conflict with any named expressions.
+             * Otherwise, mark it as an iteration variable so subsequent
+             * named expressions can check for conflicts.
+             */
+            if (val & (SYMTAB_CONSTS.DEF_GLOBAL | SYMTAB_CONSTS.DEF_NONLOCAL)) {
+                // @stu no end line and end coloffset in syntax err??
```

Comment:
correct. The traceback object doesn't care about those.

Reply:
it will soon: https://twitter.com/pyblogsal/status/1416034899639750659?s=20 I guess

Reply:
well that will be fun to implement

Reply:
![image](https://media.giphy.com/media/w89ak63KNl0nJl80ig/giphy.gif)